### PR TITLE
Support migration to self-managed CoreDNS

### PR DESCRIPTION
**Pull request by jacksmith15**
This PR aims to enable users of the module to migrate to self-managed CoreDNS (as opposed to using the EKS addon).

The [included docs](https://github.com/cookpad/terraform-aws-eks/blob/js/search-3950/support-migration-to-self-managed-coredns/docs/self_managed_coredns.md#self-managed-coredns) describe the intended process for users, as well as reasons why a user may want to manage CoreDNS themselves.

tl;dr I would like to be able to scale CoreDNS in my cluster, but am not currently able to due to the way the EKS addon will overwrite any modifications I make.

### CODE REVIEW

File: modules/cluster/variables.tf [hunk 1]
```diff
@@ -189,6 +189,18 @@ variable "critical_addons_node_group_bottlerocket_admin_container_source" {
   description = "URI of a custom admin container image"
 }
 
+variable "critical_addons_coredns_preserve" {
+  type        = string
+  default     = true
```
Comment (jacksmith15) [inline modules/cluster/variables.tf:194]: I think its safe to make this `true` by default. It reduces the number of steps required to migrate, and has no effect for users who do not want to migrate.

**Review by errm — APPROVED** [review]
LGTM ...

Please `terraform fmt`

File: modules/cluster/addons.tf [hunk 2]
```diff
@@ -46,10 +46,12 @@ resource "aws_eks_addon" "kube-proxy" {
 }
 
 resource "aws_eks_addon" "coredns" {
+  count             = var.critical_addons_coredns_enabled == true ? 1 : 0
```
Comment (errm) [inline modules/cluster/addons.tf:49]: I think this is risky actually ... please test, but I think it may force a delete and recreate on an existing cluster!

**Review by errm — CHANGES_REQUESTED** [review]

File: modules/cluster/addons.tf (hunk 2 above)
Comment (errm) [reply modules/cluster/addons.tf:49]: I am guessing we will need some instructions in UPGRADING to explain the manual state migration steps that will be needed!

Comment (jacksmith15) [reply modules/cluster/addons.tf:49]: Could I request some support on (or information about) the testing process?

The [docs mention pre-production testing](https://github.com/cookpad/terraform-aws-eks/blob/main/RELEASING.md#branching) - do we have a test cluster configured for this purpose?

**Comment by jacksmith15** [comment]
> question Did you ask AWS if they can make changes to the addon to support the behaviour you need!

Short answer: no

Long answer:

They do _already_ support the behaviour I need, but only in a roundabout manner. The EKS addon accepts a `configuration_values` parameter - which has the following parameters:

- `corefile` (configmap contents)
- `nodeSelector`
- `replicaCount`
- `resources`
- `computeType`

<details><summary><b>Configuration values schema</b></summary>

```json
{
  "$ref": "#/definitions/Coredns",
  "$schema": "http://json-schema.org/draft-06/schema#",
  "definitions": {
    "Coredns": {
      "additionalProperties": false,
      "properties": {
        "computeType": {
          "type": "string"
        },
        "corefile": {
          "description": "Entire corefile contents to use with installation",
          "type": "string"
        },
        "nodeSelector": {
          "additionalProperties": {
            "type": "string"
          },
          "type": "object"
        },
        "replicaCount": {
          "type": "integer"
        },
        "resources": {
          "$ref": "#/definitions/Resources"
        }
      },
      "title": "Coredns",
      "type": "object"
    },
    "Limits": {
      "additionalProperties": false,
      "properties": {
        "cpu": {
          "type": "string"
        },
        "memory": {
          "type": "string"
        }
      },
      "title": "Limits",
      "type": "object"
    },
    "Resources": {
      "additionalProperties": false,
      "properties": {
        "limits": {
          "$ref": "#/definitions/Limits"
        },
        "requests": {
          "$ref": "#/definitions/Limits"
        }
      },
      "title": "Resources",
      "type": "object"
    }
  }
}
```

</details>

In theory the `replicaCount` and `resources` fields are sufficient for my particular purposes. The reason I proposed the current approach is because my goal is to be able to easily respond to an outage cause by unusually high DNS load. In that case, I believe the time-to-recover will be lower if the process for tweaking vertical and horizontal scale of CoreDNS is familiar to the responder.

This is a loosely held opinion though - I'd be happy to instead expose the `configuration_values` option in this Terraform module if you think its more appropriate.

**Comment by jacksmith15** [comment]
Thanks for prompt review btw!

**Comment by ettiee** [comment]
> In theory the replicaCount and resources fields are sufficient for my particular purposes. The reason I proposed the current approach is because my goal is to be able to easily respond to an outage cause by unusually high DNS load. In that case, I believe the time-to-recover will be lower if the process for tweaking vertical and horizontal scale of CoreDNS is familiar to the responder (i.e. modify a k8s manifest).

> This is a loosely held opinion though - I'd be happy to instead expose the configuration_values option in this Terraform module if you think its more appropriate.

I'm considering exposing `configuration_values` for the EBS CSI addon, so if we make this a standard pattern it may become more familiar to the teams. And still the added benefit of not having to manage a kubernetes manifest / helm release for things provided as EKS addons

**Comment by jacksmith15** [comment]
I think it probably makes sense to expose the configuration values in this case :+1: 

I'll prepare a PR which exposes this for each of the relevant addons

**Comment by ettiee** [comment]
> I think it probably makes sense to expose the configuration values in this case 👍
> 
> I'll prepare a PR which exposes this for each of the relevant addons

Sounds good! I'll probably change my approach then and go forward with your change.
you'll need to upgrade the addon to
addon_version = "v1.10.0-eksbuild.1" 
for EBS CSI driver to support this, and some version for the other addons too I expect. 
When you've got your PR ready we can cherry pick onto a release for 1.22
